Source/WebCore/page/SecurityOriginData.h

```diff
@@ -36,49 +36,88 @@ namespace WebCore {
 class Frame;
 class SecurityOrigin;
 
+enum OpaqueOriginIdentifierType { };
+using OpaqueOriginIdentifier = ObjectIdentifier<OpaqueOriginIdentifierType>;
+
 struct SecurityOriginData {
-    enum OpaqueOriginIdentifierType { };
-    using OpaqueOriginIdentifier = ProcessQualified<ObjectIdentifier<OpaqueOriginIdentifierType>>;
-    using MarkableOpaqueOriginIdentifier = Markable<OpaqueOriginIdentifier, OpaqueOriginIdentifier::MarkableTraits>;
+    struct Tuple {
+        String protocol;
+        String host;
+        std::optional<uint16_t> port;
+        bool operator==(const Tuple& other) const { return protocol == other.protocol && host == other.host && port == other.port; }
+        Tuple isolatedCopy() const { return { protocol.isolatedCopy(), host.isolatedCopy(), port }; }
+    };
 
     SecurityOriginData() = default;
     SecurityOriginData(const String& protocol, const String& host, std::optional<uint16_t> port)
-        : protocol(protocol)
-        , host(host)
-        , port(port)
+        : data { Tuple { protocol, host, port } }
     {
         RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(!isHashTableDeletedValue());
     }
-    explicit SecurityOriginData(OpaqueOriginIdentifier opaqueOriginIdentifier)
-        : protocol(emptyString())
-        , host(emptyString())
-        , opaqueOriginIdentifier(opaqueOriginIdentifier)
-    {
-    }
+    explicit SecurityOriginData(ProcessQualified<OpaqueOriginIdentifier> opaqueOriginIdentifier)
+        : data(opaqueOriginIdentifier) { }
+    explicit SecurityOriginData(std::variant<Tuple, ProcessQualified<OpaqueOriginIdentifier>>&& data)
+        : data(WTFMove(data)) { }
     SecurityOriginData(WTF::HashTableDeletedValueType)
-        : protocol(WTF::HashTableDeletedValue)
-    {
-    }
+        : data { Tuple { WTF::HashTableDeletedValue, { }, { } } } { }
     
     WEBCORE_EXPORT static SecurityOriginData fromFrame(Frame*);
     WEBCORE_EXPORT static SecurityOriginData fromURL(const URL&);
     WEBCORE_EXPORT static SecurityOriginData fromURLWithoutStrictOpaqueness(const URL&);
 
     static SecurityOriginData createOpaque()
     {
-        return SecurityOriginData { OpaqueOriginIdentifier::generateThreadSafe() };
+        return SecurityOriginData { ProcessQualified<OpaqueOriginIdentifier>::generateThreadSafe() };
     }
 
     WEBCORE_EXPORT Ref<SecurityOrigin> securityOrigin() const;
 
     // FIXME <rdar://9018386>: We should be sending more state across the wire than just the protocol,
     // host, and port.
 
-    String protocol;
-    String host;
-    std::optional<uint16_t> port;
-    MarkableOpaqueOriginIdentifier opaqueOriginIdentifier;
+    std::variant<Tuple, ProcessQualified<OpaqueOriginIdentifier>> data;
```

Comment:
Can this be private and be named `m_data` now? I don't think outsiders should be interacting with this variant directly anymore.